### PR TITLE
New: added gRPC streams support for benchmark - 5

### PR DESCRIPTION
**Pull request by jronak**
## Aim
Support gRPC streams in Yab

## Changes
* Add benchmark handling for stream requests which works with the existing benchmarking parameters
* Reuses the initial stream request dispatcher for benchmarking
* Refactor benchmark handling into different files `benchmark_unary.go` and `benchmark_stream.go` for handlings it's respective benchmark type. 
* Stream benchmark works alongside warmup request (ad-hoc request) by recording the stream requests (via STDIN or passed on CLI) sent during the warmup and reusing the same during the Benchmarks

## How to test?
Download [grpc-stream-test-server.zip](https://github.com/yarpc/yab/files/5826809/grpc-stream-test-server.zip), follow the `readme.md` to run the test server and dispatch streaming yab queries

### Bidi Stream
```
yab  --max-requests=10000 --max-duration=1s --timeout=1m -F testproto --service "test-server" --method uber.yarpc.encoding.protobuf.Test/Stream -p grpc://localhost:55555   -r '{"value": "1"}{"value": "2"}'
// Warmup response
{
  "value": "1"
}

{
  "value": "12"
}

Benchmark parameters:
  CPUs:            12
  Connections:     24
  Concurrency:     1
  Max requests:    10000
  Max duration:    1s
  Max RPS:         0
Latencies:
  0.5000: 922.41µs
  0.9000: 2.495403ms
  0.9500: 3.514855ms
  0.9900: 22.468952ms
  0.9990: 40.554458ms
  0.9995: 40.882234ms
  1.0000: 42.377369ms
Elapsed time (seconds):   0.74
Total requests:           10000
RPS:                      13558.80
```
#### STDIN mode
```
yab  --max-requests=10000 --max-duration=1s --timeout=1m -F testproto --service "test-server" --method uber.yarpc.encoding.protobuf.Test/Stream -p grpc://localhost:55555   -r '-'
{"value":"1"} // STDIN request-1
{
  "value": "1"
}

{"value":"2"} // STDIN request-2
{
  "value": "12"
}

Benchmark parameters:
  CPUs:            12
  Connections:     24
  Concurrency:     1
  Max requests:    10000
  Max duration:    1s
  Max RPS:         0
Latencies:
  0.5000: 625.964µs
  0.9000: 1.353507ms
  0.9500: 1.899145ms
  0.9900: 16.519147ms
  0.9990: 23.410829ms
  0.9995: 23.508979ms
  1.0000: 23.947096ms
Elapsed time (seconds):   0.43
Total requests:           10000
RPS:                      23259.32
```

#### Without Warmup
```
yab --max-requests=10000 --max-duration=1s --timeout=1m -F testproto --service "test-server" --method uber.yarpc.encoding.protobuf.Test/Stream -p grpc://localhost:55555   -r '{"value": "1"}{"value": "2"}' --warmup=0
Benchmark parameters:
  CPUs:            12
  Connections:     24
  Concurrency:     1
  Max requests:    10000
  Max duration:    1s
  Max RPS:         0
Latencies:
  0.5000: 696.727µs
  0.9000: 1.639929ms
  0.9500: 2.154043ms
  0.9900: 17.015733ms
  0.9990: 27.704704ms
  0.9995: 27.858184ms
  1.0000: 28.193725ms
Elapsed time (seconds):   0.48
Total requests:           10000
RPS:                      20980.48
```

### Server Stream
```
yab  --max-requests=10000 --max-duration=1s --timeout=1m -F testproto --service "test-server" --method uber.yarpc.encoding.protobuf.Test/ServerStream -p grpc://localhost:55555   -r '{"value": "2"}'
{
  "value": "2"
}

Benchmark parameters:
  CPUs:            12
  Connections:     24
  Concurrency:     1
  Max requests:    10000
  Max duration:    1s
  Max RPS:         0
Latencies:
  0.5000: 606.604µs
  0.9000: 1.262325ms
  0.9500: 1.879435ms
  0.9900: 19.652758ms
  0.9990: 22.332575ms
  0.9995: 22.619429ms
  1.0000: 22.860896ms
Elapsed time (seconds):   0.44
Total requests:           10000
RPS:                      22716.25
```

### Client Stream
```
yab --max-requests=10000 --max-duration=1s --timeout=1m -F testproto --service "test-server" --method uber.yarpc.encoding.protobuf.Test/ClientStream -p grpc://localhost:55555   -r '{"value": "a"}{"value": "b"}'
{
  "value": "2"
}

Benchmark parameters:
  CPUs:            12
  Connections:     24
  Concurrency:     1
  Max requests:    10000
  Max duration:    1s
  Max RPS:         0
Latencies:
  0.5000: 867.434µs
  0.9000: 1.94393ms
  0.9500: 2.652286ms
  0.9900: 13.673861ms
  0.9990: 21.272389ms
  0.9995: 21.932361ms
  1.0000: 23.618049ms
Elapsed time (seconds):   0.55
Total requests:           10000
RPS:                      18123.10
```

#### STDIN mode

```
yab --max-requests=10000 --max-duration=1s --timeout=1m -F testproto --service "test-server" --method uber.yarpc.encoding.protobuf.Test/ClientStream -p grpc://localhost:55555   -r '-'
{"value": "a"} // STDIN request-1
{"value": "b"} // STDIN request-2
{
  "value": "2"
}

Benchmark parameters:
  CPUs:            12
  Connections:     24
  Concurrency:     1
  Max requests:    10000
  Max duration:    1s
  Max RPS:         0
Latencies:
  0.5000: 783.716µs
  0.9000: 1.671911ms
  0.9500: 2.270289ms
  0.9900: 10.897964ms
  0.9990: 16.759355ms
  0.9995: 16.998659ms
  1.0000: 18.837755ms
Elapsed time (seconds):   0.46
Total requests:           10000
RPS:                      21630.18

```

### CODE REVIEW

**Comment by jronak** [comment]
@Dogild is suggesting to hold off this PR until #314, #315, #316, and #317 are merged. Might have to close and re-open this PR to merge into a different branch

File: bench_method.go [hunk 1]
```diff
@@ -39,6 +40,9 @@ type peerTransport struct {
 type benchmarkMethod struct {
 	serializer encoding.Serializer
 	req        *transport.Request
+
+	streamRequest         *transport.StreamRequest
+	streamRequestMessages [][]byte
```
Comment (Dogild) [inline bench_method.go:45]: Hum, `benchmarkMethod` is now responsible of doing 2 things, unary request and streaming request. Is there any chance to decouple this by using an interface? For instance we could have:

```
type benchmarker interface {
WarmTransports(n int, tOpts TransportOptions, resolved resolvedProtocolEncoding, warmupRequests int) ([]peerTransport, error)
Call(t transport.Transport) (time.Duration, error)
}
```

and `type struct benchmarkUnaryMethod` and `type struct benchmarkStreamingMethod`

we might have a bit of duplicated code from the two structs thoughts. But I believe that future maintenance of the code will be easier with this, we already have a bunch of `if else` in the code, the interface will definitely help for this. 

what do you think?

File: handler.go [hunk 2]
```diff
@@ -108,11 +112,29 @@ func (r requestHandler) handleStreamRequest() {
...
+		// read the request messages explicitly as there was no initial request
+		for {
+			_, err := requestSupplier()
+			if err == io.EOF {
+				break
+			}
+			if err != nil {
+				r.out.Fatalf("%v\n", err)
+			}
+		}
```
Comment (Dogild) [inline handler.go:129]: Why do we need this? can we simply gives `requestSupplier` to `benchmarkMethod` ?

File: bench_method.go (hunk 1 above)
Comment (jronak) [reply bench_method.go:45]: I agree, this makes more sense. Will work on this

File: handler.go (hunk 2 above)
Comment (jronak) [reply handler.go:129]: benchmarkMethod is shared between multiple go-routines, the requests from decoder can be read only once hence this can't be shared directly. Also, requestSupplier isn't safe to be accessed in parallel. That's why the requests are accumulated and shared as a slice of request bytes

Comment (Dogild) [reply handler.go:129]: I see, is there any chance the benchmarkMethod struct takes care of this? since we will have a one type for streaming and one for unary, the streaming one should be able to do this no?

Comment (jronak) [reply handler.go:137]: This wouldn't work cleanly as benchmark could be coupled with a warm-up request. If there is a warm-up request, then request reader would have hit EOF meaning we cannot pass it to benchmarker anymore.

Comment (Dogild) [reply handler.go:137]: I see, this logic is a bit hard to follow and maintain IMO. The issues is that the code to read values from the streamer is split into 3 different parts (here, line:92 of this file, and in `benchmark_streaming.go`).

What do you think about the following proposals:

Instead of having the function `nextBodyFn := func() ([]byte, error) {}` (see https://github.com/yarpc/yab/pull/317/files#diff-0eb779b9e49d8e44b0f36923fdb8d87d5ee024f886eefc45deec4ec88380a087R91), we add a new function in handler.go:

```
func nextStreamingBodyFunction (reader, opts) func() ([]byte, error) {
   if opts == "interactive" {
      return func() ([]byte, error) {
         // read with streamMsgReader.NextBody() until eof 
      } 
   }

   // not interactive, read all of the input and create an array of streamRequestMessages[]

    // return the function
    return func() ([]byte, error) {
       return m.streamRequestMessages[streamRequestIdx]
    }
}
``` 

By having this, you would only need to call the method `nextStreamingBodyFunction` once in the handler. You could then give it to the benchmark one.

File: benchmark_streaming.go [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+	"time"
+
+	"github.com/yarpc/yab/encoding"
+	"github.com/yarpc/yab/transport"
+)
+
+// benchmarkStreamMethod benchmarks stream requests
+type benchmarkStreamMethod struct {
+	serializer encoding.Serializer
+	req        *transport.Request
```
[the remark this answers is not in the document]
Comment (jronak) [reply benchmark_streaming.go:34]: `req` is used for `Method` function, i'll try to remove it as part of the comment below

File: handler.go (hunk 2 above)
Comment (jronak) [reply handler.go:164]: I agree, the logic was spread through the streaming handler. I have split request handler and response handler into separate functions. This must simplify the logic
https://github.com/yarpc/yab/blob/3a22bcdb6f295ad5895ea197dc91d7f516e7244d/handler.go#L140

File: benchmark_stream.go [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+	start := time.Now()
+	err := makeStreamRequest(t, m.streamRequest, m.serializer, m.streamRequestSupplier(), responseHandlerFn)
+	duration := time.Since(start)
+
+	if err != nil {
+		return duration, err
+	}
+
+	// response validation is delayed to avoid consuming benchmark time for
+	// deserialising the responses.
```
Comment (Dogild) [inline benchmark_stream.go:50]: nit: typo in deserialising

File: benchmark_stream.go [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+
+// benchmarkStreamMethod benchmarks stream requests
+type benchmarkStreamMethod struct {
+	serializer            encoding.Serializer
+	streamRequest         *transport.StreamRequest
+	streamRequestMessages [][]byte
+}
+
+func (m benchmarkStreamMethod) Call(t transport.Transport) (time.Duration, error) {
+	responseHandlerFn, allResponsesFn := streamResponseRecorder()
```
Comment (Dogild) [inline benchmark_stream.go:39]: Let's use a buffering channel for the response instead of allResponsesFn. This sill simplify the code in streamResponseRecorder.

File: benchmark_stream.go [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+
+		return req, nil
+	}
+
+	return nextBodyFn
+}
+
+// streamResponseRecorder returns a stream response handler which records
+// all the responses. Recorded responses is returned by all responses function.
+func streamResponseRecorder() (streamResponseHandlerFn, func() [][]byte) {
```
Comment (Dogild) [inline benchmark_stream.go:87]: Let's add `benchmarkStreamMethod` as receiver here

File: bench_method_test.go [hunk 7]
```diff
@@ -301,10 +303,92 @@ func TestBenchmarkMethodWarmTransportsError(t *testing.T) {
 	}
 }
 
-func TestBenchmarkMethodHealth(t *testing.T) {
-	m := benchmarkMethodForROpts(t, RequestOptions{
-		Encoding: encoding.Thrift,
-		Health:   true,
-	}, transport.TChannel)
-	assert.Equal(t, "Meta::health", m.Method())
```
Comment (Dogild) [inline bench_method_test.go:309]: Why did you remove this test?

File: bench_method_test.go [hunk 8]
```diff
@@ -301,10 +303,92 @@ func TestBenchmarkMethodWarmTransportsError(t *testing.T) {
...
+				CallerName:  "test",
+				Peers:       []string{"grpc://" + lis.String()},
+			}, _resolvedGrpcProto, 1)
+			require.NoError(t, err)
+
+			assert.Equal(t, tt.expectStreams, svc.streamsOpened.Load())
+			assert.Equal(t, tt.expectSentMessage, svc.streamMessagesSent.Load())
+			assert.Equal(t, tt.expectReceiveMessage, svc.streamMessagesReceived.Load())
+		})
+	}
```
Comment (Dogild) [inline bench_method_test.go:393]: You should also add a unittest for the method `Call` of `benchmarkStreamMethod`

File: benchmark.go [hunk 9]
```diff
@@ -123,7 +140,7 @@ func runWorker(t transport.Transport, m benchmarkMethod, s *benchmarkState, run
 	}
 }
 
-func runBenchmark(out output, logger *zap.Logger, allOpts Options, resolved resolvedProtocolEncoding, m benchmarkMethod) {
+func runBenchmark(out output, logger *zap.Logger, allOpts Options, resolved resolvedProtocolEncoding, methodName string, m benchmarker) {
```
Comment (Dogild) [inline benchmark.go:143]: I remember in another comment (my bad) I did to do this `methodName` things here. But I actually think we can do better here.

Since benchmarker has a request, we should be able to get the request from the benchmarker. What about having this interface for benchmarker:

```
type benchmarker interface {
	// WarmTransports creates n transports and warms it up by dispatching specified
	// number of warm-up requests.
	WarmTransports(n int, tOpts TransportOptions, resolved resolvedProtocolEncoding, warmupRequests int) ([]peerTransport, error)

	// Call dispatches a request on the provided transport.
	Call(transport.Transport) (time.Duration, error)
	
	Request() *transport.Request
}
```

note: your original proposal was with a method `Method()`.

File: benchmark.go [hunk 10]
```diff
@@ -280,6 +297,7 @@ func outputPlaintext(out output, latencyValues map[float64]time.Duration, summar
 	out.Printf("Elapsed time (seconds):   %.2f\n", summary.ElapsedTimeSeconds)
 	out.Printf("Total requests:           %v\n", summary.TotalRequests)
 	out.Printf("RPS:                      %.2f\n", summary.RPS)
+	// TODO: add sent/received stream messages summary
```
Comment (Dogild) [inline benchmark.go:300]: Can elaborate why this was not done in this PR? This sounds like the critical feature for benchmark no?

File: handler.go [hunk 11]
```diff
@@ -88,13 +92,24 @@ func (r requestHandler) handleStreamRequest() {
 		r.out.Fatalf("Failed while preparing the request: %v\n", err)
 	}
 
+	nextBodyFn, allRequestsFn := streamRequestSupplier(streamMsgReader)
```
Comment (Dogild) [inline handler.go:95]: Can you add a comment explaining how `streamRequestSupplier` will work with the warmup and benchmark and stdin?

**Review by Dogild — COMMENTED** [review]
Please update the summary of the PR, you should add:

- Example with yab using this new feature
- Explaining the refactoring around the benchmark.go files
- How warmup is working with benchmark and warmup

File: benchmark_stream.go [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+	serializer            encoding.Serializer
+	streamRequest         *transport.StreamRequest
+	streamRequestMessages [][]byte
+}
+
+func (m benchmarkStreamMethod) Call(t transport.Transport) (time.Duration, error) {
+	responseHandlerFn, allResponsesFn := m.streamResponseRecorder()
+
+	start := time.Now()
+	err := makeStreamRequest(t, m.streamRequest, m.serializer, m.streamRequestSupplier(), responseHandlerFn)
```
Comment (Dogild) [inline benchmark_stream.go:42]: As talked offline, instead of giving function to `makeStreamRequest`, `makeStreamRequest` should ask for an interface

File: benchmark.go (hunk 10 above)
Comment (jronak) [reply benchmark.go:300]: It isn't critical in terms of functioning, but it does have some additional information such as total number of stream messages sent/received. The reason of not working on the change in this PR was to avoid any I/O changes apart from core stream benchmarking. This will be worked upon in a followup PR

File: bench_method_test.go (hunk 7 above)
Comment (jronak) [reply bench_method_test.go:309]: Benchmark method no longer exposes `Method`, so this test wouldn't be valid anymore

File: benchmark.go (hunk 9 above)
Comment (jronak) [reply benchmark.go:143]: IMO we can keep pass the method name directly in the argument, exposing transport.Request just for extracting method name seems like an overkill to me

**Review by Dogild — COMMENTED** [review]
This PR does 2 things, one refactoring of func type to interface and addition of the streaming benchmark feature.

Can we separate this into 2 PRs, this will decrease the complexity of the PRs and help reviewers.

**Review by Dogild — COMMENTED** [review]
Can we have the refactoring of benchmarker outside of the PR please? It will help to reduce number of changes in the PR and help reviewer, a lot of changes are not needed for the streaming feature.

File: integration_test.go [hunk 13]
```diff
@@ -453,6 +470,198 @@ func setupGRPCServer(t *testing.T, svc *simpleService) (net.Addr, *grpc.Server)
...
+					WarmupRequests: 1,
+					Connections:    2,
+					Concurrency:    2,
+				},
+			},
+			wantRes:              `Total requests:           100`,
+			returnOutput:         []simple.Foo{{Test: 1}},
+			expectedInput:        []simple.Foo{{Test: 1}, {Test: 2}, {Test: -1}, {Test: 10}},
+			expectStreams:        103,
+			expectReceiveMessage: 412,
```
Comment (Dogild) [inline integration_test.go:537]: How did you get 412? should it be 404? `100*len(expectedInput_) + 1*len(expectedInput)`. 
Same for `expectStreams`, why 103? (I probably missed something)

Please add a comment for those numbers how they are calculated, it will help to maintain this code

Comment (jronak) [reply integration_test.go:537]: benchmark -> `number of benchmark requests` * `len(expectedInput)` (100 * 4) = 400
warmup -> `warmup request` * `len(expectedInput)` (1 * 4) = 4
benchmark connection warmup -> `number of connections in benchmark` *  `len(expectedInput)` (2 * 4) = 8

Total: 412

Sure i'll add a comment

File: benchmark_stream_test.go [hunk 14]
```diff
@@ -0,0 +1,255 @@
...
+		expectSentMessage    int32
+		expectReceiveMessage int32
+	}{
+		{
+			name:                 "client stream success",
+			procedure:            "Bar::ClientStream",
+			method:               "Bar/ClientStream",
+			num:                  10,
+			requests:             [][]byte{nil, nil},
+			returnOutput:         []simple.Foo{{}},
```
[the remark this answers is not in the document]
Comment (jronak) [reply benchmark_stream_test.go:56]: discussed this offline, `returnOutput` is needed as it sends zero value struct in response

File: benchmark_stream_test.go [hunk 15]
```diff
@@ -0,0 +1,255 @@
...
+			expectReceiveMessage: 20,
+		},
+		{
+			name:                 "bidirectional stream success",
+			procedure:            "Bar::ServerStream",
+			method:               "Bar/ServerStream",
+			num:                  10,
+			requests:             [][]byte{nil},
+			returnOutput:         []simple.Foo{{}},
+			expectedInput:        []simple.Foo{{}},
```
[the remark this answers is not in the document]
Comment (jronak) [reply benchmark_stream_test.go:81]: Same as https://github.com/yarpc/yab/pull/319#discussion_r583397017

File: benchmark_stream_test.go [hunk 16]
```diff
@@ -0,0 +1,255 @@
...
+		expectStreams        int32
+		expectSentMessage    int32
+		expectReceiveMessage int32
+	}{
+		{
+			name:                 "client stream success",
+			procedure:            "Bar::ClientStream",
+			method:               "Bar/ClientStream",
+			requests:             [][]byte{nil, nil},
+			returnOutput:         []simple.Foo{{}},
```
[the remark this answers is not in the document]
Comment (jronak) [reply benchmark_stream_test.go:150]: same as https://github.com/yarpc/yab/pull/319#discussion_r583397017

File: benchmark_stream_test.go [hunk 17]
```diff
@@ -0,0 +1,255 @@
...
+			expectSentMessage:    2,
+			expectReceiveMessage: 2,
+		},
+		{
+			name:                 "bidirectional stream success",
+			procedure:            "Bar::ServerStream",
+			method:               "Bar/ServerStream",
+			requests:             [][]byte{nil},
+			returnOutput:         []simple.Foo{{}},
+			expectedInput:        []simple.Foo{{}},
```
[the remark this answers is not in the document]
Comment (jronak) [reply benchmark_stream_test.go:173]: same as https://github.com/yarpc/yab/pull/319#discussion_r583397017

File: benchmark_stream_test.go [hunk 18]
```diff
@@ -0,0 +1,255 @@
...
+		name                 string
+		procedure            string
+		method               string
+		num                  int
+		requests             [][]byte
+		returnOutput         []simple.Foo
+		expectedInput        []simple.Foo
+		expectStreams        int32
+		expectSentMessage    int32
+		expectReceiveMessage int32
```
[the remark this answers is not in the document]
Comment (jronak) [reply benchmark_stream_test.go:48]: Changed it to `expectedServerSentMessages` and `expectedServerReceviedMessages`

File: benchmark_stream_test.go [hunk 19]
```diff
@@ -0,0 +1,255 @@
...
+
+func TestBenchmarkMethodWarmTransportGRPCStreams(t *testing.T) {
+	tests := []struct {
+		name                 string
+		procedure            string
+		method               string
+		num                  int
+		requests             [][]byte
+		returnOutput         []simple.Foo
+		expectedInput        []simple.Foo
```
[the remark this answers is not in the document]
Comment (jronak) [reply benchmark_stream_test.go:45]: There are multiple different tests in Yab which use word `expected`, I believe it must be okay to use `expected` in tests?

https://github.com/yarpc/yab/blob/dace80fcc1439a7a161713a4f7b801ec738d702d/main_test.go#L723
https://github.com/yarpc/yab/blob/dace80fcc1439a7a161713a4f7b801ec738d702d/main_test.go#L858

File: handler.go [hunk 20]
```diff
@@ -358,6 +370,17 @@ func (s *streamIOInitializer) HandleResponse(body []byte) error {
 	return nil
 }
 
+// allRequests returns all the requests from the stream reader.
+func (s *streamIOInitializer) allRequests() [][]byte {
+	for !s.eofReached {
+		if _, err := s.NextRequest(); err != nil && err != io.EOF {
+			s.out.Fatalf("%v\n", err)
```
[the remark this answers is not in the document]
Comment (jronak) [reply handler.go]: Since allRequests was being invoked directly handleStreamRequest, it made sense to handle fatal directly as it has access to out.Fatal. Anyway, I have moved Fatal out of allRequest :-)

File: handler_test.go [hunk 21]
```diff
@@ -107,4 +109,36 @@ func TestStreamRequestRecorder(t *testing.T) {
...
 		require.Error(t, streamIO.HandleResponse([]byte("a:b")))
 	})
+
+	t.Run("all request failure", func(t *testing.T) {
+		var errBuf bytes.Buffer
+		out := testOutput{
+			fatalf: func(format string, args ...interface{}) {
+				errBuf.WriteString(fmt.Sprintf(format, args...))
+			},
+		}
```
Comment (Dogild) [inline handler_test.go:119]: `testOutput` is not used in this test anymore

File: handler.go [hunk 22]
```diff
@@ -339,6 +358,8 @@ func (s *streamIOInitializer) NextRequest() ([]byte, error) {
 		return nil, fmt.Errorf("Failed while reading stream input: %v", err)
 	}
 
+	s.streamRequests = append(s.streamRequests, msg)
+
```
Comment (Dogild) [inline handler.go:362]: nit: extra breakline

File: handler.go [hunk 23]
```diff
@@ -234,6 +247,7 @@ func makeBidiStream(ctx context.Context, stream *yarpctransport.ClientStream, st
 				// Cancel the context to unblock the routine waiting on receiving
 				// stream messages.
 				cancel()
+
```
Comment (Dogild) [inline handler.go:250]: nit: extra breakline

**Review by Dogild — APPROVED** [review]
Please fix the last comments!

File: benchmark_stream.go [hunk 24]
```diff
@@ -0,0 +1,93 @@
...
+
+	"github.com/yarpc/yab/encoding"
+	"github.com/yarpc/yab/transport"
+)
+
+// benchmarkStreamMethod benchmarks stream requests.
+type benchmarkStreamMethod struct {
+	serializer            encoding.Serializer
+	streamRequest         *transport.StreamRequest
+	streamRequestMessages [][]byte
```
Comment (AllenLuUber) [inline benchmark_stream.go:35]: not for this PR: If we throw in a `type payload []byte` and this would look like `[]payload`. Do you think it will help with readability?

File: benchmark_stream.go [hunk 25]
```diff
@@ -0,0 +1,93 @@
...
+func (b *streamIOBenchmark) HandleResponse(res []byte) error {
+	b.streamResponses = append(b.streamResponses, res)
+	return nil
+}
+
+func newStreamIOBenchmark(streamRequests [][]byte) *streamIOBenchmark {
+	return &streamIOBenchmark{
+		streamRequests: streamRequests,
+	}
+}
```
Comment (Dogild) [inline benchmark_stream.go:93]: nit: the declaration/implementation of this constructor should be below `type streamIOBenchmark struct` and above `func (b *streamIOBenchmark) NextRequest() ([]byte, error)`

**Review by Dogild — APPROVED** [review]

File: benchmark_stream_test.go [hunk 26]
```diff
@@ -0,0 +1,255 @@
...
+
+			transports, err := warmTransports(bench, tt.numConnections, TransportOptions{
+				ServiceName: "foo",
+				CallerName:  "test",
+				Peers:       []string{"grpc://" + lis.String()},
+			}, _resolvedGrpcProto, 1)
+			require.NoError(t, err)
+
+			for i, transport := range transports {
+				assert.NotNil(t, transport, "transports[%d] must not be nil", i)
```
Comment (AllenLuUber) [inline benchmark_stream_test.go:123]: nit: can we use a better indicator than index here to uniquely identify the nil transport? like name or something

File: benchmark_stream_test.go [hunk 27]
```diff
@@ -0,0 +1,255 @@
...
+			name:                                 "client stream success",
+			procedure:                            "Bar::ClientStream",
+			method:                               "Bar/ClientStream",
+			numConnections:                       10,
+			requests:                             [][]byte{nil, nil},
+			returnOutput:                         []simple.Foo{{}},
+			expectedInput:                        []simple.Foo{{}, {}},
+			expectedStreamsOpened:                10,
+			expectedServerReceivedStreamMessages: 20,
+			expectedServerSentStreamMessages:     10,
```
Comment (AllenLuUber) [inline benchmark_stream_test.go:60]: can you add some comments for the case? Took me some time to understand those abstract numbers :)
```
expectedServerReceivedStreamMessages: 20, // 10 client streams, each of them send 2 streams
expectedServerSentStreamMessages: 10, // but server responds a single message for each client stream
```

File: benchmark_stream_test.go [hunk 28]
```diff
@@ -0,0 +1,255 @@
...
+		expectedServerReceivedStreamMessages int32
+	}{
+		{
+			name:                                 "client stream success",
+			procedure:                            "Bar::ClientStream",
+			method:                               "Bar/ClientStream",
+			requests:                             [][]byte{nil, nil},
+			returnOutput:                         []simple.Foo{{}},
+			expectedInput:                        []simple.Foo{{}, {}},
+			expectedStreamsOpened:                1,
```
Comment (AllenLuUber) [inline benchmark_stream_test.go:152]: same here. a little comment after those numbers will be really helpful for future maintenance purpose.

**Review by AllenLuUber — APPROVED** [review]
LGTM. Just request more comments in the test for different cases.

File: benchmark_stream.go (hunk 24 above)
Comment (jronak) [reply benchmark_stream.go:35]: This makes sense, i'll pull up another PR with this additional type

File: benchmark_stream_test.go (hunk 26 above)
Comment (jronak) [reply benchmark_stream_test.go:123]: warmTransports creates the set of transports for the given endpoint, if the transport is nil I don't think we have any identifier except it's index?
